langpuppet/gapi.go

```diff
@@ -0,0 +1,290 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package langpuppet
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/gapi"
+	"github.com/purpleidea/mgmt/pgraph"
+	"github.com/purpleidea/mgmt/util"
+
+	"github.com/purpleidea/mgmt/lang"
+	"github.com/purpleidea/mgmt/puppet"
+
+	errwrap "github.com/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+const (
+	// Name is the name of this frontend.
+	Name = "langpuppet"
+)
+
+func init() {
+	gapi.Register(Name, func() gapi.GAPI { return &GAPI{} }) // register
+}
+
+// GAPI implements the main langpuppet GAPI interface.
+type GAPI struct {
+	InputURI string
+
+	langGAPI   gapi.GAPI
+	puppetGAPI gapi.GAPI
+
+	currentLangGraph   *pgraph.Graph
+	currentPuppetGraph *pgraph.Graph
+
+	data        gapi.Data
+	initialized bool
+	closeChan   chan struct{}
+	wg          sync.WaitGroup // sync group for tunnel go routines
+}
+
+// Cli takes a cli.Context, and returns our GAPI if activated. All arguments
+// should take the prefix of the registered name. On activation, if there are
+// any validation problems, you should return an error. If this was not
+// activated, then you should return a nil GAPI and a nil error.
+func (obj *GAPI) Cli(c *cli.Context, fs engine.Fs) (*gapi.Deploy, error) {
+	if s := c.String(Name); c.IsSet(Name) {
```

Comment:
You can inline this test's body. I didn't in other code and I forget why (probably didn't notice) but I fixed that in an upcoming feature branch.

Reply:
I think this has become a fair bit nicer in the current revision. Closing.